index.js

```diff
@@ -1,76 +1,119 @@
-// Send HTTP requests to Jira
 const processRow = row => {
 
-    // Q: What about versions? Do we need to update on all versions?
+  const pageId = row["Page ID"];
+  const title = row["Page Title"];
+  const metadata = row["text_val"];
 
-    const pageId = row["Page ID"];
-    const title = row["Page Title"];
-    const metadata = row["text_val"];
+  const issueKeys = metadata.match(ISSUE_KEY_REGEX);
+  log(INFO, `[PAGE-ID: ${pageId}] Keys: ${issueKeys}`);
 
-    const issueKeys = metadata.match(ISSUE_KEY_REGEX);
-    console.log("Keys: ", issueKeys);
+  const body = {
+    globalId: `appId=${appId}&pageId=${pageId}`,
+    application: {
+      type: "com.atlassian.confluence",
+      name: "Confluence"
+    },
+    relationship: "mentioned in",
+    object: {
+      url: `${confluenceEndpoint}/pages/viewpage.action?pageId=${pageId}`,
+      title
+    }
+  };
 
-    const body = {
-        globalId: `appId=${appId}&pageId=${pageId}`,
-        application: {
-            type: "com.atlassian.confluence",
-            name: "Confluence"
-        },
-        relationship: "mentioned in",
-        object: {
-            url: `${confluenceEndpoint}/pages/viewpage.action?pageId=${pageId}`,
-            title
-        }
-    };
+  issueKeys.forEach(issueKey => {
+    const requestUrl = `${jiraEndpoint}/rest/api/latest/issue/${issueKey}/remotelink`;
 
-    console.log(body);
+    if (isDryRun) {
+      log(INFO, `[DRY-RUN] Will update ${requestUrl}`);
+      return;
+    }
 
-    issueKeys.forEach(issueKey => {
-        const requestUrl = `${jiraEndpoint}/rest/api/latest/issue/${issueKey}/remotelink`;
-        console.log(requestUrl);
-    });
+    const authBuffer = Buffer.from(usernameAndPassword, 'utf-8');
+    requests.push(
+      axios.post(requestUrl, body, {
+        headers: {
+          'Authorization': `Basic ${authBuffer.toString('base64')}`,
+          'Content-Type': 'application/json'
+        }
+      })
+        .then(result => {
+          result.status === 201 ? createdLinksCount++ : preexistingLinksCount++;
+          log(INFO, `Update success: ${requestUrl} => ${result.status} ${JSON.stringify(result.data)}`);
+        })
+        .catch(error => {
+          errCount++
+          log(ERROR, JSON.stringify(body.object));
+          if (error.response) {
+            log(ERROR, `[ERROR] ${requestUrl} => ${error.response.status}`);
+          }
+        })
+    )
+  });
 };
 
 fs.createReadStream(path.resolve(__dirname, csvFilename))
-    .pipe(csv.parse({ headers: true }))
-    .on("error", error => console.error(error))
-    .on("data", processRow)
-    .on("end", rowCount => console.log(`Parsed ${rowCount} rows`));
+  .pipe(csv.parse({headers: true}))
+  .on("error", console.error)
+  .on("data", processRow)
+  .on("end", async rowCount => {
+    log(INFO, `Parsed ${rowCount} rows`);
+    await Promise.allSettled(requests);
+    logSummary(rowCount, requests.length, createdLinksCount, preexistingLinksCount, errCount, isDryRun);
```

Comment:
Q: If there are 100+ requests in the `requests` list, would this trigger all of them at the same time?

Reply:
Not exactly. Each request will be processed as soon as it's ready and node can allocate resources to it. To break it down:

1. The axios request is created and assigned callbacks.
2. The request (with its callbacks) is added to the `requests` array.
3. As requests are sent and their responses become available, their callbacks are triggered (while other requests are still being queued).
4. Once all requests have been queued (some will have been completed), line 117 will await for all the requests to be settled before printing the summary.

In short, requests are processed as soon as they are ready, they don't actually wait until here to start triggering.

Reply:
Thanks for the explanation, my only concern is the number of HTTP requests being sent out at one time - do you see a scenario where a large number of HTTP requests would overload the server?

Reply:
I think that the server will be okay with these lightweight requests as long as they're not executed during high traffic when the server is already taxed. Honestly, if I had a lot of time, I would have introduced batching to be extra safe but it's just a nice to have.
For reference, I ran 600+ requests on my dockerised Jira server which is running on my modestly-speced machine together with Confluence, Postgres and pgAdmin and the server did ok to handle it.
I am more concerned about the protections that may be present to prevent DoS which why I added a note about that in the README.md.